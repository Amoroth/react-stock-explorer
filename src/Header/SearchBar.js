import React, { Component } from 'react'
import { createFilter } from 'react-search-input'
import { Link } from 'react-router-dom'

import Overlay from '../shared/Overlay'
import styles from './Header.module.css'

class SearchBar extends Component {
  constructor(props) {
    super(props)
    this.state = {
      input: '',
      symbols: [],
      searchResults: false,
      searchBoxInput: false,
    }
    this.inputRef = React.createRef();
  }

  componentDidMount() {
    fetch('https://api.iextrading.com/1.0/ref-data/symbols?filter=symbol,name').then((res) => {
      return res.json()
    }).then((json) => {
      this.setState({ symbols: json, searchBoxInput: window.innerWidth > 800 })
    })
    this.setState({ symbols: [
      {name: 'Alphabet Inc.', symbol: 'googl'},
      {name: 'Microsoft Corporation', symbol: 'msft'},
      {name: 'International Buissnes Machines Corporation', symbol: 'ibm'},
      {name: 'Twitter Inc.', symbol: 'twtr'}
    ],
    searchBoxInput: window.innerWidth > 800 })
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.state.input !== prevState.input) {
      return true
    }
    return false
  }

  handleChange = (e) => {
    this.setState({ input: e.target.value })
  }

  onClickFill = (term) => {
    this.setState({ input: term })
  }

  showSearchResults = () => {
    this.setState({ searchResults: true })
  }

  hideSearchResults = () => {
    setTimeout(() => this.setState({ searchResults: false, searchBoxInput: window.innerWidth > 800 }), 100)
  }

  showSearchBoxInput = () => {
    this.setState({ searchBoxInput: true }, () => {
      this.inputRef.current.focus()
    })
  }

  render() {
    const filteredCmps = this.state.symbols.filter(
      createFilter(this.state.input, ['name', 'symbol'])
    )

    return (
      <div style={{display: 'inline'}}>
        {this.state.searchBoxInput ? 
        <input
          className={styles['search-box']}
          onChange={this.handleChange}
          value={this.state.input}
          onFocus={this.showSearchResults}
          onBlur={this.hideSearchResults}
          ref={this.inputRef}
        /> : 
        <button onClick={this.showSearchBoxInput} className={styles['search-box-icon']}>
          <i className="material-icons">search</i>
        </button>}

        <div className={styles['search-box-results']}>
          {this.state.searchResults ? filteredCmps.slice(0, 4).map((cmp) => {
            return (
              <Link
                className={styles['search-box-result']}
                key={cmp.symbol}
                onClick={() => this.onClickFill(cmp.name)}
                to={`stock?cmp=${cmp.symbol}`}>
                {cmp.name}
                <span>{cmp.symbol}</span>
              </Link>
            )}) : null}
        </div>
        {this.state.searchResults && window.innerWidth < 800 ? <Overlay /> : null}
      </div>
    )
  }
}

export default SearchBar