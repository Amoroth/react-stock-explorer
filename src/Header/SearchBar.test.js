import React from 'react'
import { shallow } from 'enzyme'

import SearchBar from './SearchBar'

describe('<SearchBar />', () => {
  let props
  let wrapper

  beforeEach(() => {
    props = { history: { push: jest.fn() } }
    wrapper = shallow(<SearchBar.WrappedComponent {...props} />)
  })

  it('renders', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('fetches search companies', () => {
    expect(fetch).toBeCalledWith('https://api.iextrading.com/1.0/ref-data/symbols?filter=symbol,name')
  })

  it('renders input on desktop', () => {
    window.innerWidth = 1000
    window.dispatchEvent(new Event('resize'))

    expect(wrapper.find('.search-box').exists()).toBe(true)
  })

  it('input shows search results on focus', () => {
    window.innerWidth = 1000
    window.dispatchEvent(new Event('resize'))
    wrapper.setState({ symbols: [{ name: 'Company', symbol: 'Symbol' }] })

    wrapper.find('.search-box').simulate('focus')
    expect(wrapper.state('searchResults')).toBe(true)
    expect(wrapper.find('Link').exists()).toBe(true)
  })

  it('input hides search results on blur', (done) => {
    window.innerWidth = 1000
    window.dispatchEvent(new Event('resize'))
    wrapper.setState({
      searchResults: true,
      symbols: [{ name: 'Company', symbol: 'Symbol' }],
    })

    wrapper.find('.search-box').simulate('blur')
    setTimeout(() => {
      expect(wrapper.state('searchResults')).toBe(false)
      expect(wrapper.find('Link').exists()).toBe(false)
      done()
    }, 200)
  })

  it('input can be controled with keyboard', () => {
    window.innerWidth = 1000
    window.dispatchEvent(new Event('resize'))
    wrapper.setState({
      searchResults: true,
      symbols: [{ name: 'Company', symbol: 'SYM' }, { name: 'Other', symbol: 'OTH' }],
    })

    wrapper.find('.search-box').simulate('keyup', { keyCode: 40 })
    wrapper.find('.search-box').simulate('keyup', { keyCode: 40 })
    wrapper.find('.search-box').simulate('keyup', { keyCode: 40 })
    expect(wrapper.state('inputCursor')).toBe(1)
    wrapper.find('.search-box').simulate('keyup', { keyCode: 38 })
    expect(wrapper.state('inputCursor')).toBe(0)
    wrapper.find('.search-box').simulate('keyup', { keyCode: 38 })
    expect(wrapper.state('inputCursor')).toBe(0)
  })

  it('input can be selected with enter', () => {
    window.innerWidth = 1000
    window.dispatchEvent(new Event('resize'))
    wrapper.setState({
      searchResults: true,
      symbols: [{ name: 'Company', symbol: 'SYM' }, { name: 'Other', symbol: 'OTH' }],
    })
    const blurFn = jest.fn()
    wrapper.instance().inputRef = { current: { blur: blurFn } }

    wrapper.find('.search-box').simulate('keyup', { keyCode: 40 })
    wrapper.find('.search-box').simulate('keyup', { keyCode: 13 })
    expect(blurFn).toBeCalled()
    expect(wrapper.state('input')).toBe('Company')
    expect(props.history.push).toBeCalledWith('stock?cmp=SYM')
  })

  it('search results render at most 4 links', () => {
    wrapper.setState({
      searchResults: true,
      symbols: [
        { name: 'Company', symbol: 'SYM' },
        { name: 'Other', symbol: 'OTH' },
        { name: 'Other2', symbol: 'OT2' },
        { name: 'Other3', symbol: 'OT3' },
        { name: 'Other4', symbol: 'OT4' },
        { name: 'Other5', symbol: 'OT5' },
      ],
    })

    expect(wrapper.find('Link').length).toBe(4)
  })

  it('click on link fills input', () => {
    wrapper.setState({
      searchResults: true,
      symbols: [{ name: 'Company', symbol: 'SYM' }],
    })

    wrapper.find('Link').first().simulate('click')
    expect(wrapper.state('input')).toBe('Company')
  })

  it('showes button on mobile', () => {
    expect(wrapper.find('.search-box').exists()).toBe(false)
    expect(wrapper.find('.search-box-icon').exists()).toBe(true)
  })

  it('search button on mobile showes search-box', () => {
    wrapper.instance().inputRef = { current: { focus: jest.fn() } }

    wrapper.find('.search-box-icon').simulate('click')
    expect(wrapper.state('searchBoxInput')).toBe(true)
  })

  it('search button on mobile focuses on input', () => {
    const focusFn = jest.fn()
    wrapper.instance().inputRef = { current: { focus: focusFn } }

    wrapper.find('.search-box-icon').simulate('click')
    expect(focusFn).toBeCalled()
  })

  it('on resize switches input with button', () => {
    expect(wrapper.find('.search-box-icon').exists()).toBe(true)
    expect(wrapper.find('.search-box').exists()).toBe(false)

    window.innerWidth = 1000
    window.dispatchEvent(new Event('resize'))

    expect(wrapper.find('.search-box-icon').exists()).toBe(false)
    expect(wrapper.find('.search-box').exists()).toBe(true)
  })
})
