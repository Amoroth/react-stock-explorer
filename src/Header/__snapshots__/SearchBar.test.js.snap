// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> renders 1`] = `
<div
  style={
    Object {
      "display": "inline",
    }
  }
>
  <button
    className="search-box-icon"
    onClick={[Function]}
    type="button"
  >
    <i
      className="material-icons"
    >
      search
    </i>
  </button>
  <div
    className="search-box-results"
  />
</div>
`;
